prober/derp.go

```diff
@@ -577,6 +580,196 @@ func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc
 	return nil
 }
 
+// derpProbeBandwidthMu ensures that derpProbeBandwidthTUN doesn't run concurrently.
+// This is necessary to avoid conflicts trying to create the TUN device.
+var derpProbeBandwidthMu sync.Mutex
+
+// derpProbeBandwidthTUN takes two DERP clients (fromc and toc) connected to two
+// DERP servers (from and to) and sends a test payload of a given size from one
+// to another over TUN.
+func derpProbeBandwidthTUN(ctx context.Context, transferTime *expvar.Float, from, to *tailcfg.DERPNode, fromc, toc *derphttp.Client, size int64) error {
+	derpProbeBandwidthMu.Lock()
+	defer derpProbeBandwidthMu.Unlock()
+
+	// Temporarily set up a TUN device with which to simulate a real client TCP connection
+	// tunneling over DERP.
+	dev, _, err := tstun.New(log.Printf, "utun9")
+	if err != nil {
+		return fmt.Errorf("failed to create TUN device: %w", err)
+	}
+	defer dev.Close()
+	mtu, err := dev.MTU()
+	if err != nil {
+		return fmt.Errorf("failed to get TUN MTU: %w", err)
+	}
+
+	r, err := router.New(log.Printf, dev, nil, nil)
```

Comment:
What do we need the `router` for? I'm assuming to number the TUN?

Reply:
I'm using it to configure routes so that we can send packets via the TUN.

Reply:
Numbering the TUN will give us a connected route (unless Darwin differs here for TUN). If it's serving the numbering purpose, that's fine.

Reply:
> Numbering the TUN

What does this mean?

What I observe on Mac is that if I don't configure a router at all, I can't bind a socket to `172.16.0.1` and the network interface shown by `ifconfig` has no IP address.

```
utun9: flags=8051<UP,POINTOPOINT,RUNNING,MULTICAST> mtu 1280
```

If however I at least set a local address like this

```
r.Set(&router.Config{
    LocalAddrs: []netip.Prefix{netip.MustParsePrefix("172.16.0.1/32")},
})
```

Then I can bind a socket to `172.16.0.1` and the interface actually has an IP address.

```
utun9: flags=8051<UP,POINTOPOINT,RUNNING,MULTICAST> mtu 1280
	inet 172.16.0.1 --> 172.16.0.1 netmask 0xffffffff
```